### PR TITLE
unix: use a heap for timers

### PR DESCRIPTION
**Pull request by saghul**
> Replace the red-black tree with a heap.  The most common operation that
> libuv performs on timers is looking up the first timer to expire.  With
> a red-black tree, that operation is O(log n).  With a heap, it's O(1).

This work was done by @bnoordhuis a while ago. I talked with him and the code is ready to be merged. While there might be room for more optimizations when using millions of timers, the common use cases are already better with this.

This patch is a squash and rebase of https://github.com/joyent/libuv/pull/928

@bnoordhuis: does heap-inl.h need to be listed in Makefile.am and uv.gyp? It didn't complain here, but we seem to list all internal include files.

### CODE REVIEW

File: src/unix/timer.c [hunk 1]
```diff
@@ -20,35 +20,41 @@
...
-    return -1;
-  if (a->timeout > b->timeout)
     return 1;
-  /*
-   *  compare start_id when both has the same timeout. start_id is
-   *  allocated with loop->timer_counter in uv_timer_start().
+  if (b->timeout < a->timeout)
+    return 0;
+
+  /* Ccompare start_id when both has the same timeout. start_id is
```
Comment (bnoordhuis) [inline src/unix/timer.c]: Typo. :-)

**Comment by bnoordhuis** [comment]
> does heap-inl.h need to be listed in Makefile.am and uv.gyp?

Yes, it would be good to have it listed.

**Comment by saghul** [comment]
Fixed the typo and added the file to Makefile.am and uv.gyp.

**Comment by retep998** [comment]
Speak of the devil. I was just benchmarking running millions of timers, firing every second.
Using the latest official release, I can get to about 1.6 million timers before it can't keep up.
Using the version in this pull request lets me get to about 2.5 million before it slows down.

**Comment by saghul** [comment]
@retep998 sweet, thanks for sharing the results!

File: src/heap-inl.h [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#ifndef UV_SRC_HEAP_H_
+#define UV_SRC_HEAP_H_
+
+#include <stddef.h>  /* NULL */
+
+#if defined(__GNUC__)
+# define HEAP_EXPORT(declaration) __attribute__((unused)) static declaration
```
Comment (indutny) [inline src/heap-inl.h:22]: Is there any point for `declaration` arg to be here, except style? I'd rather not use it.

Comment (saghul) [inline src/heap-inl.h:22]: How would you put it?

```
# define HEAP_EXPORT __attribute__((unused)) static
```

Comment (indutny) [inline src/heap-inl.h:22]: Yep, I think we are doing `UV_EXTERN` the same way.

File: src/heap-inl.h [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+  unsigned int k;
+
+  newnode->left = NULL;
+  newnode->right = NULL;
+  newnode->parent = NULL;
+
+  /* Calculate the path from the root to the insertion point.  This is a min
+   * heap so we always insert at the left-most free node of the bottom row.
+   */
+  path = 0;
```
Comment (indutny) [inline src/heap-inl.h:122]: Is this extra step really required?

File: src/heap-inl.h [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    if (path & 1)
+      child = &(*child)->right;
+    else
+      child = &(*child)->left;
+    path >>= 1;
+    k -= 1;
+  }
+
+  /* Insert the new node. */
+  newnode->parent = *parent;
```
Comment (indutny) [inline src/heap-inl.h:139]: Am I right that root will be it's own parent?

Comment (indutny) [inline src/heap-inl.h:139]: nvm, I was wrong.

File: src/heap-inl.h (hunk 3 above)
Comment (indutny) [inline src/heap-inl.h:122]: Ah, I see...

File: src/heap-inl.h (hunk 2 above)
Comment (saghul) [inline src/heap-inl.h:22]: I don't mind changing it, but I trust the original author's criteria ;-)

**Comment by indutny** [comment]
LGTM

**Comment by saghul** [comment]
Landed, thanks!

**Pull request merged**